### PR TITLE
SI-8267 Avoid existentials after polymorphic overload resolution

### PR DESCRIPTION
**Pull request by retronym**
... which can be introduced by `memberType` for methods
with methods dependent on class type parameters.

Usually these are skolemized out of existence, but when
we carry through multiple overloaded alternatives and deferred
the call to `memberType`, this wasn't happening.

This commit calls `adapt` on function tree after polymorphic
overload resolution has run. This is analagous to `doTypedApply`:

``` scala
inferMethodAlternative(fun, undetparams, argTpes, pt)
doTypedApply(tree, adapt(fun, mode.forFunMode, WildcardType), args1, mode, pt)
```

This REPL transcript might further illustrate the problem.

```
scala> class Bippy { trait Foo[A] }
defined class Bippy

scala> final class RichBippy[C <: Bippy with Singleton](val c1: C) {
     |   def g[A](x: A)(ev: c1.Foo[A]): Int = 2
     | }
defined class RichBippy

scala> :power
** Power User mode enabled - BEEP WHIR GYVE **
** :phase has been set to 'typer'.          **
** scala.tools.nsc._ has been imported      **
** global._, definitions._ also imported    **
** Try  :help, :vals, power.<tab>           **

scala> val g = typeOf[RichBippy[_]].member(TermName("g"))
g: $r.intp.global.Symbol = method g

scala> val c = new Bippy
c: Bippy = Bippy@92e2c93

scala> val memberType = typeOf[RichBippy[c.type]].memberType(g)
memberType: $r.intp.global.Type = ([A](x: A)(ev: _7.c1.Foo[A])Int) forSome { val _7: RichBippy[c.type] }

scala> val tree = Ident(TermName("dummy")).setType(memberType)
tree: $r.intp.global.Ident = dummy

scala> def adaptExprWildcard(t: Tree) = {
     |   type HasAdapt = { def adapt(t: Tree, mode: Int, pt: Type, original: Tree) }
     |   typer.asInstanceOf[HasAdapt].adapt(t, Mode.EXPRmode.bits, WildcardType, EmptyTree)
     | }
warning: there was one feature warning; re-run with -feature for details
adaptExprWildcard: (t: $r.intp.global.Tree)Unit

scala> adaptExprWildcard(tree)

scala> tree.tpe
res1: $r.intp.global.Type = [A](x: A)(ev: _7.c1.Foo[A])Int

scala> memberType.skolemizeExistential
res2: $r.intp.global.Type = [A](x: A)(ev: _7.c1.Foo[A])Int
```

Review by @lrytz / @adriaanm

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala [hunk 1]
```diff
@@ -1448,7 +1448,7 @@ trait Infer extends Checkable {
       }
       matchingLength.alternatives match {
         case Nil        => fail()
-        case alt :: Nil => finish(alt, pre memberType alt)
+        case alt :: Nil => finish(alt, (pre memberType alt).skolemizeExistential(context.owner, tree))
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala]: Why do you also need to `skolemizeExistential` here? Could you add a comment?

**Comment by lrytz** [comment]
I think I got it other than the question above.

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala (hunk 1 above)
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala]: Oops, that's left over from my first attempt to fix this in which I copy/pasted the relevant line from `adapt` into a few places here. I'll remove it post haste.

**Comment by retronym** [comment]
I guess the other point to debate here is what butterfly effect the extra call to `adapt` might have, and whether that would be better targeted to 2.12.x.

I was originally quite concerned about:

``` scala
        case PolyType(tparams, restpe) if mode.inNone(TAPPmode | PATTERNmode) && !context.inTypeConstructorAllowed => // (3)
          // assert((mode & HKmode) == 0) //@M a PolyType in HKmode represents an anonymous type function,
          // we're in HKmode since a higher-kinded type is expected --> hence, don't implicitly apply it to type params!
          // ticket #2197 triggered turning the assert into a guard
          // I guess this assert wasn't violated before because type aliases weren't expanded as eagerly
          //  (the only way to get a PolyType for an anonymous type function is by normalisation, which applies eta-expansion)
          // -- are we sure we want to expand aliases this early?
          // -- what caused this change in behaviour??
          val tparams1 = cloneSymbols(tparams)
          val tree1 = (
            if (tree.isType) tree
            else TypeApply(tree, tparams1 map (tparam => TypeTree(tparam.tpeHK) setPos tree.pos.focus)) setPos tree.pos
          )
          context.undetparams ++= tparams1
          notifyUndetparamsAdded(tparams1)
          adapt(tree1 setType restpe.substSym(tparams, tparams1), mode, pt, original)
```

But then I realised we can't get there in `TAPPmode`.

**Comment by retronym** [comment]
In any case, perhaps we want to get a second opinion on this one from @adriaanm or @odersky before we merge.

**Comment by retronym** [comment]
I should also mention that this code path is only for explicit type applications. When type arguments are inferred we get into `doTypedApply` which appears to already be doing the right thing.

**Comment by lrytz** [comment]
Much agree to have a few more eyes on it. To me it looks good, but I would not be surprised if one of the Types wizards points out some overlooked interplay.

**Comment by lrytz** [comment]
Much agree to have a few more eyes on it. To me it looks good, but I would not be surprised if one of the Types wizards points out some overlooked interplay.

**Comment by adriaanm** [comment]
Why not call skolemizeExistential directly?

**Comment by retronym** [comment]
That's what I did in my first attempt, but I felt that reusing `adapt` reduces duplication and is more consistent with the typechecking of applications with inferred type args.

**Comment by adriaanm** [comment]
Maybe we should factor out the duplication from adapt and use it directly? In general, I think it's better to avoid calling adapt. I'll poke around a bit and look at the archives first.

**Comment by retronym** [comment]
> consistent with the typechecking of applications with inferred type args.

.. and with typechecking of explicit type applications to non-overloaded methods.

**Comment by adriaanm** [comment]
I still feel we're addressing the problem in the wrong place, though I'm not sure how to do better yet.
I think you identified the real culprit correctly in the bug report:

```
        case tp if mode.typingExprNotLhs && isExistentialType(tp) =>
          adapt(tree setType tp.dealias.skolemizeExistential(context.owner, tree), mode, pt, original)
```

How come we don't reach the case "skolemize existential" (aka open the existential) after overloading resolution, a few cases higher in adapt? 

```
        case OverloadedType(pre, alts) if !mode.inFunMode => // (1)
          inferExprAlternative(tree, pt)
          adapt(tree, mode, pt, original)
```

**Comment by retronym** [comment]
We're in `FUNmode` (we've got argument list on the way), which is why we can't pick an overload in `adapt`.

I just checked when that path through `adapt` is actually triggered. Nothing in `pos` hits it, but compiling the library finds this code pattern:

``` scala
object Overload {
  def apply[T](c:Any) = 0
  def apply[T]() = 0

  Overload[Int]
}
```

``` scala
adapt(
  tree = Overload#8490.apply#16613[Int#1909] // .tpe = ()Int#1909 <and> (c#16383: Any#741)Int#1909
  pt = ?
)
```

**Comment by retronym** [comment]
Which is to say, in this test case overload resolution is done in `inferPolyAlternatives` using the provided type arguments to filter the alternatives.

**Comment by adriaanm** [comment]
Okay, that made it click for me. Could we create an adaptOverloaded method or marker to denote all the spots where overload resolution is done and the resulting type must thus be further adapted? Maybe a comment marker suffices.

I would like some polish for the comment & commit message. Maybe render the PR discussion in the comment? As it stands, its train of thought is hard to follow.

Also, the commit message has a weird fragment: "for methods with methods"  (They say you like methods!? Ok! I'll put some methods in your methods!)

**Comment by retronym** [comment]
I've rewritten the commit comment and added the marker method, `adaptAfterOverloadResolution`.

**Comment by adriaanm** [comment]
LGTM! :shipit:
